src/concurrency/timestamp_ordering_transaction_manager.cpp

```diff
@@ -397,83 +184,286 @@ bool TimestampOrderingTransactionManager::AcquireOwnership(
 // will release the write lock anyway.
 void TimestampOrderingTransactionManager::YieldOwnership(
-    UNUSED_ATTRIBUTE Transaction *const current_txn, const oid_t &tile_group_id,
+    UNUSED_ATTRIBUTE Transaction *const current_txn, 
+    const storage::TileGroupHeader *const tile_group_header,
     const oid_t &tuple_id) {
-  auto &manager = catalog::Manager::GetInstance();
-  auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
   PL_ASSERT(IsOwner(current_txn, tile_group_header, tuple_id));
   tile_group_header->SetTransactionId(tuple_id, INITIAL_TXN_ID);
 }
 
 bool TimestampOrderingTransactionManager::PerformRead(
-    Transaction *const current_txn, const ItemPointer &location,
+    Transaction *const current_txn, const ItemPointer &read_location,
     bool acquire_ownership) {
-  if (current_txn->IsDeclaredReadOnly() == true) {
-    // Ignore read validation for all readonly transactions
+
+  ItemPointer location = read_location;
+
+  //////////////////////////////////////////////////////////
+  //// handle READ_ONLY
+  //////////////////////////////////////////////////////////
+  if (current_txn->GetIsolationLevel() == IsolationLevelType::READ_ONLY) {
+    // do not update read set for read-only transactions.
     return true;
-  }
+  } // end READ ONLY
 
-  oid_t tile_group_id = location.block;
-  oid_t tuple_id = location.offset;
+  //////////////////////////////////////////////////////////
+  //// handle SNAPSHOT
+  //////////////////////////////////////////////////////////
 
-  LOG_TRACE("PerformRead (%u, %u)\n", location.block, location.offset);
-  auto &manager = catalog::Manager::GetInstance();
-  auto tile_group = manager.GetTileGroup(tile_group_id);
-  auto tile_group_header = tile_group->GetHeader();
-
-  // Check if it's select for update before we check the ownership and modify
-  // the
-  // last reader tid
-  if (acquire_ownership == true &&
-      IsOwner(current_txn, tile_group_header, tuple_id) == false) {
-    // Acquire ownership if we haven't
-    if (IsOwnable(current_txn, tile_group_header, tuple_id) == false) {
-      // Can not own
-      return false;
-    }
-    if (AcquireOwnership(current_txn, tile_group_header, tuple_id) == false) {
-      // Can not acquire ownership
-      return false;
+  // TODO: what if we want to read a version that we write? 
+  else if (current_txn->GetIsolationLevel() == IsolationLevelType::SNAPSHOT) {
+  
+    oid_t tile_group_id = location.block;
+    oid_t tuple_id = location.offset;
+
+    LOG_TRACE("PerformRead (%u, %u)\n", location.block, location.offset);
+    auto &manager = catalog::Manager::GetInstance();
+    auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
+
+    // Check if it's select for update before we check the ownership 
+    // and modify the last reader cid
+    if (acquire_ownership == true) {
+
+      // get the latest version of this tuple.
+      location = *(tile_group_header->GetIndirection(location.offset));
```

Comment:
it's `select_for_update`. We read the snapshot version and update the latest version. We need to acquire write locks to prevent write-write conflicts. See Section VIII.B in the paper: Multi-version Concurrency via Timestamp Range Conflict Management.